ecommerce/api.py

```diff
@@ -167,85 +169,81 @@ def get_valid_coupon_versions(product, user, auto_only=False, code=None):
+            payment_version__amount=decimal.Decimal(1)
+        )
 
-        # filter by redemption counts
-        for coupon_version in query:
-            redemptions_global = (
-                CouponRedemption.objects.select_related("coupon_version", "order")
-                .filter(coupon_version=coupon_version)
-                .filter(order__status=Order.FULFILLED)
-            )
-            redemptions_user = redemptions_global.filter(order__purchaser=user)
-            if (
-                coupon_version.payment_version.max_redemptions
-                > redemptions_global.count()
-                and coupon_version.payment_version.max_redemptions_per_user
-                > redemptions_user.count()
-            ):
-                valid_coupons.append(coupon_version)
-        return sorted(
-            valid_coupons, key=lambda x: x.payment_version.amount, reverse=True
+    if company is not None:
+        coupon_version_subquery = coupon_version_subquery.filter(
+            payment_version__company=company
         )
 
+    coupon_version_subquery = coupon_version_subquery.order_by(
+        "coupon", "-created_on"
+    ).distinct("coupon")
 
-def best_coupon_for_basket(basket, auto_only=False, code=None):
-    """
-    Get the best eligible coupon for the basket.
-    Assumes that the basket only contains one item/product.
+    # Exclude versions with too many redemptions or active dates outside of today.
+    query = (
+        CouponVersion.objects.select_related("coupon", "payment_version")
+        .filter(pk__in=Subquery(coupon_version_subquery.values_list("pk", flat=True)))
+        .filter(
+            Q(payment_version__expiration_date__gte=now)
+            | Q(payment_version__expiration_date__isnull=True)
+        )
+        .filter(
+            Q(payment_version__activation_date__lte=now)
+            | Q(payment_version__activation_date__isnull=True)
+        )
+        .annotate(
+            user_redemptions=(
+                Count(
+                    "couponredemption",
+                    filter=(
+                        Q(couponredemption__order__purchaser=user)
+                        & Q(couponredemption__order__status=Order.FULFILLED)
+                    ),
+                )
+            )
+        )
+        .annotate(
+            global_redemptions=(
+                Count(
+                    "couponredemption",
+                    filter=(Q(couponredemption__order__status=Order.FULFILLED)),
+                )
+            )
+        )
+        .exclude(user_redemptions__gte=F("payment_version__max_redemptions_per_user"))
+        .exclude(global_redemptions__gte=F("payment_version__max_redemptions"))
+    )
 
-    Args:
-        basket (Basket): the basket Object
-        auto_only (bool): Only retrieve `automatic` Coupons
-        code (str): A coupon code to filter by
+    if auto_only:
+        query = query.filter(payment_version__automatic=True)
```

Comment:
Since this is a filter that is conditionally applied by a kwarg, it seems like we should group this with the other conditional filters. Maybe below the company filter?